Add ability to negate dynamic queues

kcrayon opened the pull request:
Pretty handy feature.

Lets you set !@queue to exclude queues listed in @queue

Commit:
Add ability to blacklist dynamic queues.

Commit:
Slightly smarther dynamic queue negation.

snmaynard committed:
Shuffle the matching queues

Commit:
Merge pull request #1 from snmaynard/patch-2

Shuffle the matching queues

kcrayon committed:
Sort view of queues in web.

wr0ngway commented:
Thanks!

wr0ngway commented on lib/resque/plugins/dynamic_queues/queues.rb:
```diff
@@ -19,7 +19,7 @@ module Queues
           queue_names = @queues.dup
 
-          return queues_without_dynamic if queue_names.grep(/(^!)|(^@)|(\*)/).size == 0
+          return queues_without_dynamic.shuffle if queue_names.grep(/(^!)|(^@)|(\*)/).size == 0
```
Why the shuffle?  That breaks the standard resque behavior of returning queues in alpha order...
If you want fair (random) queue distribution, check out resque-fairly or my plugin resque-queue-priority

wr0ngway commented on lib/resque/plugins/dynamic_queues/queues.rb:
```diff
@@ -50,7 +54,7 @@ def queues_with_dynamic
             end
           end
 
-          return matched_queues.uniq.sort
```
sort here is a shortcut by me to to preserve sorting like standard resque - Its probably more correct for me to preserve ordering instead (as I did in resque-queue-priority), but shuffling is even more an the unexpected behavior for standard resque users

wr0ngway commented on spec/queues_spec.rb:
```diff
@@ -129,6 +129,12 @@
       worker.queues.should == ["high_x", "high_y", "superhigh_z"]
     end
 
+    it "can blacklist dynamic queues" do
```
Should probably add a test to exercise the toggling nature of the  map on line 35 - i.e. when mykey is ["foo", "!bar"] then !@mykey applies ["!foo", "bar"]

wr0ngway commented on lib/resque/plugins/dynamic_queues/server.rb:
```diff
@@ -15,7 +15,7 @@ def self.registered(app)
                   'name' => k,
                   'value' => Array(v).join(", "),
-                  'expanded' => Resque::Worker.new("@#{k}").queues.join(", ")
+                  'expanded' => Resque::Worker.new("@#{k}").queues.sort.join(", ")
```
This sort not needed when queues is sorted to preserve resque contract

Commit:
Moving shuffling into branch.

Commit:
Add test for dynamic queue with negation

kcrayon commented:
You're right about the shuffling (this is something I need, but don't know an clean way to implement), I've removed it and added a test for negating dynamic queues.

wr0ngway commented:
Cool, thanks - if you use resque-fairly or resque-queue-priority to do the shuffle, just make sure you load it after this plugin due to the sort it performs right before returning the queues
